fix: Resolve AttributeError in TableWidget and improve initialization

shuoweil opened the pull request:
This PR introduces enhancements to the `TableWidget` by improving its initialization and robustness, specifically addressing an `AttributeError` related to the `_batches` attribute.
  
   Key improvements include:
   **Enhanced Initialization:** Ensures the `_batches` attribute is correctly initialized in the `TableWidget` constructor by calling `_reset_batches_for_new_page_size` before it is accessed, making the widget more reliable from creation.
   **Improved Typing:** Added explicit type hints for `_batches` to resolve MyPy errors, contributing to better code clarity and maintainability.
   **Refactored Batch Handling:** Changed the `to_pandas_batches` call to the internal  `_to_pandas_batches` to ensure the correct `PandasBatches` object is returned, satisfying type requirements and streamlining internal data handling.
   **Test Enhancements:** Added new tests to verify the widget's row count is immutable after creation and that it gracefully handles cases with invalid total rows, ensuring the stability of the new functionality.

shuoweil committed:
remove expensive len() call

shuoweil committed:
add testcase

shuoweil committed:
fix a typo

shuoweil committed:
change how row_count is updated

shuoweil committed:
testcase stil fails, need to merged in 1888

tswast commented:
Please also update the benchmarks to use the `total_rows` parameter.

shuoweil committed:
update the method of using PandasBatches.total_rows

shuoweil commented on bigframes/display/anywidget.py:
```diff
@@ -74,15 +76,18 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
+            self.row_count = self._batches.total_rows or 0
+        else:
+            # Fallback for compatibility
+            self.row_count = 0
```
You are right, I have updated my code accordingly.

shuoweil commented:
> Please also update the benchmarks to use the `total_rows` parameter.

Let's use a separate PR for this request.

shuoweil commented on bigframes/display/anywidget.py:
```diff
@@ -62,29 +66,38 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
+                self.row_count = self._batches.total_rows or 0
+            else:
+                self.row_count = 0
+            self.page_size = initial_page_size
```
A comment is added.

shuoweil committed:
change tests in read_gbq_colab

shuoweil committed:
polish comment

shuoweil committed:
fix a test

shuoweil commented:
Per checking, the failed testcase multimodal_test.py::test_multimodal_dataframe and /tmpfs/src/github/python-bigquery-dataframes/notebooks/location/regionalized.ipynb  are not introduced by my change. @tswast

tswast commented on tests/benchmark/read_gbq_colab/first_page.py:
```diff
@@ -27,8 +27,9 @@ def first_page(*, project_id, dataset_id, table_id):
-    # Get number of rows (to calculate number of pages) and the first page.
-    df.shape
+    # Use total_rows from batches directly and the first page
+    execute_result = df._block.session._executor.execute(df._block.expr, ordered=True)
+    execute_result.total_rows or 0
     next(iter(df.to_pandas_batches(page_size=PAGE_SIZE)))
```
I think this would cause the query to execute twice. Please revert. Instead, it should match as closely as possible to what happens in TableWidget.

```suggestion
    # Get number of rows (to calculate number of pages) and the first page.
    batches = df.to_pandas_batches(page_size=PAGE_SIZE)
    first_page = next(iter(batches))
    assert first_page is not None
    total_rows = batches.total_rows
    assert total_rows is not None
```

shuoweil committed:
change code and update more testcase

shuoweil committed:
remove unneeded except

shuoweil committed:
add assert for total_rows

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -160,15 +167,17 @@ def _get_next_batch(self) -> bool:
             self._cached_batches.append(batch)
             return True
-        except StopIteration:
+        except (StopIteration, ValueError) as e:
             self._all_data_loaded = True
+            if isinstance(e, ValueError):
+                # If we fail to get a batch, assume no more data is available.
+                self.row_count = 0
```
We should not hide such errors. `ValueError` indicates some bug in BigFrames or the BigQuery API. The user needs to know about those.

shuoweil committed:
get actual row_counts

shuoweil committed:
avoid two query calls

shuoweil commented on bigframes/display/anywidget.py:
```diff
@@ -160,15 +168,22 @@ def _get_next_batch(self) -> bool:
+            StopIteration,
+            google.api_core.exceptions.GoogleAPICallError,
+            TypeError,
+            ValueError,
```
I see. I remove the invalid test case then.

shuoweil committed:
remove double query when display widget

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -60,32 +63,37 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
+        execute_result = dataframe._block.session._executor.execute(
+            dataframe._block.expr,
+            ordered=True,
+            use_explicit_destination=True,
```
I don't think we want to use an explicit destination here. This would result in creating a BigQuery job every time, which is not desirable. We want to allow the faster job optional code paths.

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -60,32 +63,37 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
+        # The query issued by `to_pandas_batches()` already contains metadata
+        # about how many results there were. Use that to avoid doing an extra
+        # COUNT(*) query that `len(...)` would do.
+        self.row_count = execute_result.total_rows or 0
```
Why did you switch it to use the execute result instead of the PandasBatches object returned by `to_pandas_batches`?

shuoweil committed:
get row count directly

shuoweil committed:
restore notebook

shuoweil committed:
restore notebook change

shuoweil committed:
remove duplicated code

shuoweil commented on bigframes/display/anywidget.py:
```diff
@@ -60,32 +74,31 @@ def __init__(self, dataframe: bigframes.dataframe.DataFrame):
 
-        super().__init__()
         self._dataframe = dataframe
+        self._initializing = True
```
I used to have some difficulty with use-specified page size, thus I introduce this flag. I have removed _page_size_default but keep this flag

shuoweil committed:
minor updates

shuoweil committed:
still have zero total rows issue

shuoweil committed:
now large dataset can get the correct row counts

shuoweil committed:
benchmark change

tswast commented on bigframes/session/bq_caching_executor.py:
Please revert the changes in this file.

shuoweil replied:
This change has been reverted,   a bug is caught here. Fix in b/450057742

tswast commented on tests/benchmark/read_gbq_colab/aggregate_output.py:
```diff
@@ -25,9 +25,17 @@ def aggregate_output(*, project_id, dataset_id, table_id):
-    next(iter(df.to_pandas_batches(page_size=PAGE_SIZE)))
+    # Call the executor directly to isolate the query execution time
+    # from other DataFrame overhead for this benchmark.
+    execute_result = df._block.session._executor.execute(
```
Please remove the use of `.execute()` here and in the TableWidget implementation.

shuoweil committed:
revert a benchmark

shuoweil committed:
revert executor change

shuoweil committed:
raising a NotImplementedError when the row count is none

shuoweil committed:
change return type

shuoweil committed:
Revert accidental change of dataframe.ipynb

shuoweil committed:
remove unnecessary execution in benchmark

shuoweil committed:
remove row_count check

shuoweil committed:
remove extra execute_result

shuoweil committed:
remove unnecessary tests

tswast commented on bigframes/core/blocks.py:
```diff
@@ -99,7 +99,7 @@
 
 
 @dataclasses.dataclass
-class PandasBatches(Iterator[pd.DataFrame]):
```
Why did you remove this subclass?

shuoweil replied:
I have added it back

shuoweil committed:
Fix: Address review comments on PandasBatches and docstring

- Reinstated 'Iterator[pd.DataFrame]' inheritance for 'PandasBatches' in 'bigframes/core/blocks.py'.
- Removed internal type hint 'bigframes.core.blocks.PandasBatches:' from 'to_pandas_batches' docstring in 'bigframes/dataframe.py' to avoid exposing internal types in public documentation.

shuoweil committed:
Revert: Revert import change in read_gbq_colab benchmark

This reverts the import path for the benchmark utils to 'benchmark.utils' to address concerns about google3 imports.

shuoweil committed:
Revert: Revert unnecessary changes in read_gbq_colab benchmarks

shuoweil committed:
Remove notebooks/Untitled-2.ipynb

shuoweil committed:
Remove notebooks/multimodal/audio_transcribe_partial_ordering.ipynb

shuoweil committed:
remove unnecessary change

shuoweil committed:
revert typo

shuoweil committed:
add todo

shuoweil committed:
change docstring

shuoweil committed:
revert changes to tests/benchmark/read_gbq_colab

shuoweil committed:
merge change

tswast commented on bigframes/display/anywidget.py:
```diff
@@ -23,13 +23,14 @@
 import pandas as pd
 
 import bigframes
+from bigframes.core.blocks import PandasBatches
```
> Use import statements for packages and modules only, not for individual types, classes, or functions.

https://google.github.io/styleguide/pyguide.html#22-imports

```suggestion
from bigframes.core import blocks
```

shuoweil commented on tests/system/small/test_anywidget.py:
```diff
@@ -436,6 +455,82 @@ def test_widget_creation_should_load_css_for_rendering(table_widget):
+    """
+    Given an internal component fails to return valid execution data,
+    when the TableWidget is created, its row_count should safely fall back to 0.
+    """
```
I updated the testcase and code accordingly.

shuoweil committed:
update how we handle invalid row count

shuoweil committed:
eliminate duplated flags
